### PR TITLE
Subresource maxDepth

### PR DESCRIPTION
**Pull request by Nightbr**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1316
| License       | MIT
| Doc PR        | https://github.com/api-platform/docs/pull/351

Proper PR from original: https://github.com/api-platform/core/pull/1512

### CODE REVIEW

**Comment by Nightbr** [comment]
Hey @soyuka,

I tried to create a clean PR and rebase with `git fetch origin && git rebase origin/master` but event after I resolve all conflict it keeps create all this mess... I don't understand why this is happening :/

**Comment by soyuka** [comment]
Don't worry about rebasing we can do it later ;). 

I think that you were close from a correct rebase but then you did a `git merge` (probably from a `git pull`) that broke everything. When rebased, you need to force push because it rewrites your history.

To remove the last merge: `git reset --hard HEAD~1` then `git push -f`

**Comment by soyuka** [comment]
Arrgh I fixed it, why did you push again? ^^

**Comment by Nightbr** [comment]
humm, I think it's Gitkraken... It make automatic git pull... I'm sorry

**Comment by Nightbr** [comment]
I think its ok now 😅

**Comment by soyuka** [comment]
@Nightbr no problem haha, FYI `git pull` does `git fetch + git merge` which can be an unwanted thing.

File: src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php [hunk 1]
```diff
@@ -82,7 +82,8 @@ private function updateMetadata(ApiSubresource $annotation, PropertyMetadata $pr
         $type = $propertyMetadata->getType();
         $isCollection = $type->isCollection();
         $resourceClass = $isCollection ? $type->getCollectionValueType()->getClassName() : $type->getClassName();
+        $maxDepth = $annotation->maxDepth ? $annotation->maxDepth : null;
```
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Do we really want to turn 0 into null?

File: src/Operation/Factory/SubresourceOperationFactory.php [hunk 2]
```diff
@@ -79,6 +81,15 @@ private function computeSubresourceOperations(string $resourceClass, array &$tre
...
 
             $visiting = "$resourceClass $property $subresourceClass";
+
+            // Handle maxDepth
+            if ($rootResourceClass === $resourceClass) {
+                $maxDepth = $subresource->getMaxDepth();
+            }
+
+            if (null !== $maxDepth && $depth >= $maxDepth) {
+                continue;
```
Comment (meyerbaptiste) [inline src/Operation/Factory/SubresourceOperationFactory.php]: Should not it be `break` here?

File: src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php (hunk 1 above)
Comment (Nightbr) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: depth 0 has no sens I think. If it's 0 or null, we do not apply maxDepth. If maxDepth=1 we create one subresource operation, if maxDepth=2 we create two subresources depth, ...

Comment (soyuka) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: null is correct here yes

File: src/Operation/Factory/SubresourceOperationFactory.php (hunk 2 above)
Comment (soyuka) [reply src/Operation/Factory/SubresourceOperationFactory.php]: :+1:

Comment (Nightbr) [reply src/Operation/Factory/SubresourceOperationFactory.php]: done! ty

File: src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php (hunk 1 above)
Comment (meyerbaptiste) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Right, the behaviour should be the same but IMO we should not transform the configured value: 0 is not null (it's important in terms of extension point). What do you think about keeping 0? The behavior will remain the same with your condition in `SubresourceOperationFactory::computeSubresourceOperations()`.

Comment (Nightbr) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Good point! fixed

File: src/Metadata/Property/SubresourceMetadata.php [hunk 3]
```diff
@@ -54,4 +56,12 @@ public function withCollection(bool $collection): self
 
         return $metadata;
     }
+
+    /**
+     * @return int
```
Comment (dunglas) [inline src/Metadata/Property/SubresourceMetadata.php]: `@return int|null`

File: src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php [hunk 4]
```diff
@@ -82,7 +82,8 @@ private function updateMetadata(ApiSubresource $annotation, PropertyMetadata $pr
         $type = $propertyMetadata->getType();
         $isCollection = $type->isCollection();
         $resourceClass = $isCollection ? $type->getCollectionValueType()->getClassName() : $type->getClassName();
+        $maxDepth = $annotation->maxDepth;
```
Comment (sroze) [inline src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Shouldn't we check that `maxDepth` is `int` here?

Comment (dunglas) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Doctrine Annotation will handle that because of the PHPDock IIRC.

Comment (sroze) [reply src/Metadata/Property/Factory/AnnotationSubresourceMetadataFactory.php]: Oh, I didn't know that. Nice.

File: src/Metadata/Property/SubresourceMetadata.php (hunk 3 above)
Comment (Nightbr) [reply src/Metadata/Property/SubresourceMetadata.php]: fixed!

**Comment by soyuka** [comment]
Thanks @Nightbr !

**Comment by Nightbr** [comment]
Thanks for your help!

**Pull request merged**